File: legacy_backend/aggregators/podcast.py
"""
Podcast RSS aggregator.

This module provides an aggregator for podcast RSS feeds (iTunes/RSS 2.0).

Features:
- Detects iTunes podcast namespace
- Extracts audio enclosure URL
- Parses duration from iTunes tags
- Creates embedded audio player content
"""

import logging
from typing import Any

from pydantic import BaseModel, Field

from .base import BaseAggregator, RawArticle

logger = logging.getLogger(__name__)


class PodcastAggregatorConfig(BaseModel):
    id: str
    type: str = Field(pattern="^(managed|custom|social)$")
    name: str = Field(min_length=1)
    url: str = ""
    description: str = Field(min_length=1)
    wait_for_selector: str | None = None
    selectors_to_remove: list[str] = Field(default_factory=list)
    options: dict = Field(default_factory=dict)


def parse_duration_to_seconds(duration_str: str) -> int | None:
    """
    Parse podcast duration string to seconds.

    Supports formats:
    - HH:MM:SS (01:23:45)
    - MM:SS (23:45)
    - Seconds only (1234)
    """
    if not duration_str:
        return None

    duration_str = duration_str.strip()

    # Try seconds only
    if duration_str.isdigit():
        return int(duration_str)

    # Try HH:MM:SS or MM:SS format
    parts = duration_str.split(":")
    try:
        if len(parts) == 3:
            hours, minutes, seconds = map(int, parts)
            return hours * 3600 + minutes * 60 + seconds
        elif len(parts) == 2:
            minutes, seconds = map(int, parts)
            return minutes * 60 + seconds
    except (ValueError, TypeError):
        pass

    return None


class PodcastAggregator(BaseAggregator):
    """
    Aggregator for podcast RSS feeds (iTunes/RSS 2.0).

    Podcast RSS feeds typically include:
    - <enclosure> with audio file URL and MIME type
    - <itunes:duration> for episode length
    - <itunes:image> for episode/show artwork
    - <itunes:summary> or <description> for episode notes
    """

    id = "podcast"
    type = "custom"
    name = "Podcast"
    url = ""
    description = (
        "Aggregator for podcast RSS feeds (iTunes/RSS 2.0). "
        "Extracts audio files, duration, and show notes. "
        "Creates embedded HTML5 audio players for each episode."
    )
    selectors_to_remove = []

    def __init__(self):
        super().__init__()
        PodcastAggregatorConfig(
            id=self.id,
            type=self.type,
            name=self.name,
            url=self.url,
            description=self.description,
            wait_for_selector=self.wait_for_selector,
            selectors_to_remove=self.selectors_to_remove,
            options=self.options,
        )

    def parse_entry(self, entry: Any) -> RawArticle:
        """
        Parse podcast RSS entry with audio metadata.

        Podcast entries include:
        - enclosure: Audio file URL and type
        - itunes:duration: Episode length
        - itunes:image: Episode artwork
        """
        article = super().parse_entry(entry)

        # Extract and store podcast-specific data
        audio_url, audio_type = self._extract_enclosure(entry)
        entry["_podcast_audio_url"] = audio_url
        entry["_podcast_audio_type"] = audio_type
        entry["_podcast_duration"] = self._extract_duration(entry)
        entry["_podcast_image"] = self._extract_image(entry)

        return article

    def _extract_enclosure(self, entry: Any) -> tuple[str, str]:
        """Extract audio enclosure URL and MIME type."""
        enclosures = entry.get("enclosures", [])

        for enclosure in enclosures:
            url = enclosure.get("url", "") or enclosure.get("href", "")
            media_type = enclosure.get("type", "")

            # Look for audio types
            if media_type.startswith("audio/") or url.lower().endswith(
                (".mp3", ".m4a", ".wav", ".ogg", ".opus", ".aac")
            ):
                return url, media_type

        # Try links with enclosure rel
        for link in entry.get("links", []):
            if link.get("rel") == "enclosure":
                url = link.get("href", "")
                media_type = link.get("type", "")
                if url:
                    return url, media_type

        return "", ""

    def _extract_duration(self, entry: Any) -> int | None:
        """Extract episode duration in seconds."""
        # Try itunes:duration
        duration_str = entry.get("itunes_duration", "")
        if duration_str:
            return parse_duration_to_seconds(duration_str)

        # Try duration
        duration_str = entry.get("duration", "")
        if duration_str:
            return parse_duration_to_seconds(duration_str)

        return None

    def _extract_image(self, entry: Any) -> str:
        """Extract episode or show artwork URL."""
        # Try itunes:image (episode-specific)
        itunes_image = entry.get("itunes_image", {})
        if isinstance(itunes_image, dict):
            url = itunes_image.get("href", "")
            if url:
                return url

        # Try image (RSS standard)
        image = entry.get("image", {})
        if isinstance(image, dict):
            url = image.get("href", "") or image.get("url", "")
            if url:
                return url

        # Try media:thumbnail
        media_thumbnail = entry.get("media_thumbnail", [])
        if media_thumbnail and isinstance(media_thumbnail, list):
            return media_thumbnail[0].get("url", "")

        return ""

    def fetch_article_html(self, article: RawArticle) -> str:
        """
        Generate HTML content with embedded audio player.

        Creates:
        - Episode artwork (if available)
        - HTML5 audio player with controls
        - Download link
        - Episode description/show notes
        """
        entry = article.entry
        audio_url = entry.get("_podcast_audio_url", "")
        audio_type = entry.get("_podcast_audio_type", "audio/mpeg")
        duration = entry.get("_podcast_duration")
        image_url = entry.get("_podcast_image", "")

        # Extract description
        description = self._extract_description(entry)

        html_parts = []

        # Episode artwork
        if image_url:
            html_parts.append(
                f'<div class="podcast-artwork">'
                f'<img src="{image_url}" alt="Episode artwork" loading="lazy">'
                f"</div>"
            )

        # Audio player
        if audio_url:
            html_parts.append(
                f'<div class="podcast-player">'
                f'<audio controls preload="metadata">'
                f'<source src="{audio_url}" type="{audio_type or "audio/mpeg"}">'
                f"Your browser does not support the audio element."
                f"</audio>"
            )

            # Duration badge
            if duration:
                formatted_duration = self._format_duration(duration)
                html_parts.append(
                    f'<span class="podcast-duration">{formatted_duration}</span>'
                )

            # Download link
            html_parts.append(
                f'<a href="{audio_url}" class="podcast-download" download>'
                f'<i class="bi bi-download"></i> Download Episode'
                f"</a>"
                f"</div>"
            )

        # Episode description/show notes
        if description:
            html_parts.append('<div class="podcast-description">')
            html_parts.append("<h4>Show Notes</h4>")
            # Parse HTML or convert plain text
            if "<" in description and ">" in description:
                # Already HTML
                html_parts.append(description)
            else:
                # Plain text - convert to paragraphs
                paragraphs = description.split("\n\n")
                for para in paragraphs:
                    para = para.strip()
                    if para:
                        para = para.replace("\n", "<br>")
                        html_parts.append(f"<p>{para}</p>")
            html_parts.append("</div>")

        return "\n".join(html_parts)

    def _extract_description(self, entry: Any) -> str:
        """Extract episode description/show notes."""
        # Try content:encoded (full HTML)
        for content in entry.get("content", []):
            if content.get("type") == "text/html":
                return content.get("value", "")

        # Try itunes:summary
        summary = entry.get("itunes_summary", "")
        if summary:
            return summary

        # Try description
        summary = entry.get("summary", "") or entry.get("description", "")
        return summary

    def _format_duration(self, seconds: int) -> str:
        """Format duration in seconds to HH:MM:SS or MM:SS."""
        hours, remainder = divmod(seconds, 3600)
        minutes, secs = divmod(remainder, 60)
        if hours:
            return f"{hours}:{minutes:02d}:{secs:02d}"
        return f"{minutes}:{secs:02d}"

    def extract_content(self, article: RawArticle) -> None:
        """Content is already extracted in fetch_article_html."""
        pass

    def save_article(self, article: RawArticle, content: str) -> bool:
        """
        Save article with podcast-specific metadata.

        Stores:
        - thumbnail_url: Episode artwork
        - media_url: Audio file URL
        - duration: Episode length in seconds
        - media_type: Audio MIME type
        """
        from django.utils import timezone

        from core.models import Article

        entry = article.entry
        audio_url = entry.get("_podcast_audio_url", "")
        audio_type = entry.get("_podcast_audio_type", "")
        duration = entry.get("_podcast_duration")
        image_url = entry.get("_podcast_image", "")

        # Use current timestamp if feed is configured for it (default: True)
        if self.feed and getattr(self.feed, "use_current_timestamp", True):
            article_date = timezone.now()
        else:
            article_date = article.date

        _, created = Article.objects.update_or_create(
            url=article.url,
            defaults={
                "feed": self.feed,
                "name": article.title,
                "date": article_date,
                "content": content,
                "thumbnail_url": image_url,
                "media_url": audio_url,
                "duration": duration,
                "media_type": audio_type or "audio/mpeg",
            },
        )

        if created:
            self.logger.info(f"Created podcast episode: {article.title}")

        return created


# Module-level wrapper for compatibility
def aggregate(feed, force_refresh=False, options=None):
    """Module-level wrapper for admin interface."""
    aggregator = PodcastAggregator()
    return aggregator.aggregate(feed, force_refresh, options or {})

File: src/server/aggregators/podcast.ts
/**
 * Podcast aggregator.
 *
 * Aggregator for podcast RSS feeds (iTunes/RSS 2.0).
 *
 * Features:
 * - Detects iTunes podcast namespace
 * - Extracts audio enclosure URL
 * - Parses duration from iTunes tags
 * - Creates embedded audio player content
 */

import * as cheerio from "cheerio";

import { BaseAggregator } from "./base/aggregator";
import { fetchFeed } from "./base/fetch";
import type { RawArticle } from "./base/types";

/**
 * Parse duration string to seconds.
 */
function parseDurationToSeconds(durationStr: string): number | null {
  if (!durationStr) return null;

  durationStr = durationStr.trim();

  // Try seconds only
  if (/^\d+$/.test(durationStr)) {
    return parseInt(durationStr, 10);
  }

  // Try HH:MM:SS or MM:SS format
  const parts = durationStr.split(":");
  try {
    if (parts.length === 3) {
      const hours = parseInt(parts[0], 10);
      const minutes = parseInt(parts[1], 10);
      const seconds = parseInt(parts[2], 10);
      return hours * 3600 + minutes * 60 + seconds;
    } else if (parts.length === 2) {
      const minutes = parseInt(parts[0], 10);
      const seconds = parseInt(parts[1], 10);
      return minutes * 60 + seconds;
    }
  } catch {
    // Invalid format
  }

  return null;
}

/**
 * Format duration in seconds to HH:MM:SS or MM:SS.
 */
function formatDuration(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const remainder = seconds % 3600;
  const minutes = Math.floor(remainder / 60);
  const secs = remainder % 60;

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, "0")}:${secs.toString().padStart(2, "0")}`;
  }
  return `${minutes}:${secs.toString().padStart(2, "0")}`;
}

/**
 * Get property value from object by trying multiple possible keys.
 */
function getAnyProperty<T>(
  obj: Record<string, unknown>,
  keys: string[],
): T | null {
  for (const key of keys) {
    if (obj[key] !== undefined && obj[key] !== null) {
      return obj[key] as T;
    }
  }
  return null;
}

/**
 * Extract audio enclosure from RSS entry.
 */
function extractEnclosure(item: Record<string, unknown>): {
  url: string;
  type: string;
} {
  // Try enclosures array
  const enclosures = item["enclosures"];
  if (Array.isArray(enclosures) && enclosures.length > 0) {
    for (const enclosure of enclosures as unknown[]) {
      const enclosureObj = enclosure as {
        url?: string;
        href?: string;
        type?: string;
      };
      const url = enclosureObj.url || enclosureObj.href || "";
      const mediaType = enclosureObj.type || "";

      // Look for audio types
      if (
        mediaType.startsWith("audio/") ||
        /\.(mp3|m4a|wav|ogg|opus|aac)$/i.test(url)
      ) {
        return { url, type: mediaType || "audio/mpeg" };
      }
    }
  }

  // Try links with enclosure rel
  const links = item["links"];
  if (Array.isArray(links)) {
    for (const link of links as unknown[]) {
      const linkObj = link as { rel?: string; href?: string; type?: string };
      if (linkObj.rel === "enclosure" && linkObj.href) {
        return { url: linkObj.href, type: linkObj.type || "audio/mpeg" };
      }
    }
  }

  return { url: "", type: "" };
}

/**
 * Extract duration from RSS entry.
 */
function extractDuration(item: Record<string, unknown>): number | null {
  const keys = ["itunes_duration", "itunes:duration", "duration"];
  const durationValue = getAnyProperty<string>(item, keys);

  if (typeof durationValue === "string") {
    return parseDurationToSeconds(durationValue);
  }

  return null;
}

/**
 * Extract episode or show artwork URL.
 */
function extractImage(item: Record<string, unknown>): string {
  const keys = ["itunes_image", "itunes:image", "image"];
  const image = getAnyProperty<string | { href?: string; url?: string }>(
    item,
    keys,
  );

  if (image != null) {
    if (typeof image === "object") {
      return image.href || image.url || "";
    }
    if (typeof image === "string") {
      return image;
    }
  }

  // Try media_thumbnail
  const mediaThumbs =
    item["media_thumbnail"] || item["media:thumbnail"] || null;
  if (Array.isArray(mediaThumbs) && mediaThumbs.length > 0) {
    return (mediaThumbs[0] as { url?: string }).url || "";
  }

  return "";
}

/**
 * Extract episode description/show notes.
 */
function extractDescription(item: Record<string, unknown>): string {
  // Try content:encoded (full HTML)
  const contents = item["content"];
  if (Array.isArray(contents)) {
    for (const content of contents as unknown[]) {
      const contentObj = content as { type?: string; value?: string };
      if (contentObj.type === "text/html" && contentObj.value) {
        return contentObj.value;
      }
    }
  }

  const keys = ["content_encoded", "itunes_summary", "itunes:summary"];
  const summaryValue = getAnyProperty<string>(item, keys);
  if (typeof summaryValue === "string") return summaryValue;

  const summary = item["summary"];
  const description = item["description"];
  if (typeof summary === "string") return summary;
  if (typeof description === "string") return description;

  return "";
}

export class PodcastAggregator extends BaseAggregator {
  override readonly id = "podcast";
  override readonly type: "managed" | "custom" | "social" = "custom";
  override readonly name = "Podcast";
  override readonly url = "";
  override readonly description =
    "Podcast RSS feeds - Audio content distributed via RSS with episodes, show notes, and metadata.";

  override readonly identifierType = "url" as const;
  override readonly identifierLabel = "Podcast Feed URL";
  override readonly identifierDescription = "RSS feed URL for the podcast";
  override readonly prefillName = false;

  /**
   * Fetch RSS feed data.
   */
  protected override async fetchSourceData(
    limit?: number,
  ): Promise<import("rss-parser").Output<unknown>> {
    const startTime = Date.now();
    this.logger.info(
      {
        step: "fetchSourceData",
        subStep: "start",
        aggregator: this.id,
        feedId: this.feed?.id,
        limit,
      },
      "Fetching podcast RSS feed",
    );

    if (!this.feed) {
      throw new Error("Feed not initialized");
    }

    const feedUrl = this.feed.identifier;
    const feed = await fetchFeed(feedUrl);

    if (feed.items?.length === 0) {
      this.logger.warn(
        {
          step: "fetchSourceData",
          subStep: "complete",
          aggregator: this.id,
          feedId: this.feed?.id,
          feedUrl,
        },
        "No entries found in podcast feed",
      );
    }

    const elapsed = Date.now() - startTime;
    this.logger.info(
      {
        step: "fetchSourceData",
        subStep: "complete",
        aggregator: this.id,
        feedId: this.feed?.id,
        itemCount: feed.items?.length || 0,
        elapsed,
      },
      "Podcast RSS feed fetched",
    );

    return feed;
  }

  /**
   * Parse podcast RSS feed items to RawArticle[].
   */
  protected override async parseToRawArticles(
    sourceData: unknown,
  ): Promise<RawArticle[]> {
    const startTime = Date.now();
    this.logger.info(
      {
        step: "parseToRawArticles",
        subStep: "start",
        aggregator: this.id,
        feedId: this.feed?.id,
      },
      "Parsing podcast episodes",
    );

    const feed = sourceData as import("rss-parser").Output<unknown>;
    const items = feed.items || [];
    const articles: RawArticle[] = [];

    for (const item of items) {
      try {
        const article = await this.parseEpisode(
          item as Record<string, unknown>,
        );
        if (article) {
          articles.push(article);
        }
      } catch (error) {
        this.logger.error(
          {
            step: "parseToRawArticles",
            subStep: "parseEpisode",
            aggregator: this.id,
            feedId: this.feed?.id,
            error: error instanceof Error ? error : new Error(String(error)),
            item,
          },
          "Error processing podcast episode",
        );
      }
    }

    const elapsed = Date.now() - startTime;
    this.logger.info(
      {
        step: "parseToRawArticles",
        subStep: "complete",
        aggregator: this.id,
        feedId: this.feed?.id,
        articleCount: articles.length,
        elapsed,
      },
      "Podcast episodes parsed",
    );

    return articles;
  }

  /**
   * Parse a single podcast episode.
   */
  private async parseEpisode(
    item: Record<string, unknown>,
  ): Promise<RawArticle | null> {
    // Extract podcast-specific data
    const { url: audioUrl, type: audioType } = extractEnclosure(item);
    const duration = extractDuration(item);
    const imageUrl = extractImage(item);
    const description = extractDescription(item);

    if (!audioUrl) {
      this.logger.warn(
        {
          step: "parseToRawArticles",
          subStep: "parseEpisode",
          aggregator: this.id,
          feedId: this.feed?.id,
          title: String(item.title || "Unknown"),
        },
        "Podcast episode has no audio enclosure, skipping",
      );
      return null;
    }

    // Build HTML content with embedded audio player
    const htmlParts: string[] = [];

    // Episode artwork
    if (imageUrl) {
      htmlParts.push(
        `<div class="podcast-artwork"><img src="${imageUrl}" alt="Episode artwork" loading="lazy"></div>`,
      );
    }

    // Audio player
    htmlParts.push(
      `<div class="podcast-player"><audio controls preload="metadata"><source src="${audioUrl}" type="${audioType || "audio/mpeg"}">Your browser does not support the audio element.</audio>`,
    );

    // Duration badge
    if (duration) {
      const formattedDuration = formatDuration(duration);
      htmlParts.push(
        `<span class="podcast-duration">${formattedDuration}</span>`,
      );
    }

    // Download link
    htmlParts.push(
      `<a href="${audioUrl}" class="podcast-download" download><i class="bi bi-download"></i> Download Episode</a></div>`,
    );

    // Episode description/show notes
    if (description) {
      htmlParts.push('<div class="podcast-description">');
      htmlParts.push("<h4>Show Notes</h4>");
      htmlParts.push(this.formatEpisodeDescription(description));
      htmlParts.push("</div>");
    }

    const content = htmlParts.join("\n");

    return {
      title: String(item.title || "Untitled"),
      url: String(item.link || ""),
      published: item.pubDate ? new Date(String(item.pubDate)) : new Date(),
      content,
      summary: description,
      thumbnailUrl: imageUrl || undefined,
      mediaUrl: audioUrl,
      duration: duration || undefined,
      mediaType: audioType || "audio/mpeg",
    };
  }

  /**
   * Format episode description into HTML.
   */
  private formatEpisodeDescription(description: string): string {
    // Parse HTML or convert plain text
    if (description.includes("<") && description.includes(">")) {
      // Already HTML - sanitize it
      const $ = cheerio.load(description);
      $("script, iframe, embed, object").remove();
      return $.html();
    }

    // Plain text - convert to paragraphs
    const paragraphs = description.split("\n\n");
    const htmlParts: string[] = [];
    for (const para of paragraphs) {
      const trimmed = para.trim();
      if (trimmed) {
        const withBreaks = trimmed.replace(/\n/g, "<br>");
        htmlParts.push(`<p>${withBreaks}</p>`);
      }
    }
    return htmlParts.join("\n");
  }

  /**
   * Check if content should be fetched - podcast aggregator never fetches.
   */
  protected override shouldFetchContent(_article: RawArticle): boolean {
    // Podcast aggregator uses content from feed, never fetches from web
    return false;
  }
}
